### PR TITLE
Fix overscaling when the controller is much faster then the listener

### PR DESCRIPTION
**Pull request by nikola-jokic**
When the ephemeral runner finishes, the listener issues a scale-down patch to the ephemeral runner set. But if ephemeral runner set reacts quicker than the listener, it may re-create the ephemeral runner, and quickly remove it.
To avoid this problem,  this PR introduces the following changes:
1. Ephemeral runner cleans up its resources as soon as it is done
2. Ephemeral runner set keeps finished ephemeral runners until the next patch request is received.
3. The target count is calculated only on patch ID difference (or patchID == 0), i.e. when the listener patches the ephemeral runner set.
4. The listener issues a patch request when there is a change in the state, or in case it gets notified that the job is done.

This change also requires a change in the CRD.

### CODE REVIEW

File: apis/actions.github.com/v1alpha1/ephemeralrunnerset_types.go [hunk 1]
```diff
@@ -24,6 +24,8 @@ import (
 type EphemeralRunnerSetSpec struct {
 	// Replicas is the number of desired EphemeralRunner resources in the k8s namespace.
 	Replicas int `json:"replicas,omitempty"`
+	// PatchID is unique identifier for the patch issued by the listener app
```
Comment (rentziass) [inline apis/actions.github.com/v1alpha1/ephemeralrunnerset_types.go]:
```suggestion
	// PatchID is the unique identifier for the patch issued by the listener app
```

File: charts/gha-runner-scale-set-controller/crds/actions.github.com_ephemeralrunnersets.yaml [hunk 2]
```diff
@@ -6957,9 +6957,14 @@ spec:
                         - containers
                       type: object
                   type: object
+                patchID:
+                  description: PatchID is unique identifier for the patch issued by the listener app
```
Comment (rentziass) [inline charts/gha-runner-scale-set-controller/crds/actions.github.com_ephemeralrunnersets.yaml]:
```suggestion
                  description: PatchID is the unique identifier for the patch issued by the listener app
```

File: config/crd/bases/actions.github.com_ephemeralrunnersets.yaml [hunk 3]
```diff
@@ -6957,9 +6957,14 @@ spec:
                         - containers
                       type: object
                   type: object
+                patchID:
+                  description: PatchID is unique identifier for the patch issued by the listener app
```
Comment (rentziass) [inline config/crd/bases/actions.github.com_ephemeralrunnersets.yaml]:
```suggestion
                  description: PatchID is the unique identifier for the patch issued by the listener app
```

**Review by rentziass — DISMISSED** [review]
This looks good to me, I just left a bunch of suggestions over a type!

File: charts/gha-runner-scale-set-controller/crds/actions.github.com_ephemeralrunnersets.yaml (hunk 2 above)
Comment (nikola-jokic) [reply charts/gha-runner-scale-set-controller/crds/actions.github.com_ephemeralrunnersets.yaml]: Generated with `make manifests` :relaxed:

**Review by rentziass — APPROVED** [review]

**Pull request merged**